source/game.programme.programmelicence.bmx

```diff
@@ -1137,6 +1137,11 @@ Type TProgrammeLicence Extends TBroadcastMaterialSource {_exposeToLua="selected"
 			'all licences need to exceed the limit
 			'return GetSublicenceExceedingBroadcastLimitCount() = GetSubLicenceCount()
 
+			'TODO um selbstproduzierte Serien korrekt zu behandeln,
+			'müsste man noch prüfen, ob die Anzahl der Kinder mit
+			'der Anzahl der Episoden aus dem zugrundeliegenden Skript übereinstimmt
+			'sonst würde eine partiell abgedrehte Serie schon zurückgegeben werden
+
```

Comment:
Man koennte ueberlegen, einen Flag einzufuehren, der bei Serien/Element-mit-Kindelementen ein "Complete" markiert. Wenn das aber nur fuer diese Stelle genutzt wuerde, waere wohl ein simpler check ueber das Drehbuch der weniger invasive Einsatz.

Reply:
```blitzmax
if programmeData.productionID > 0
'alternative
'if programmeData.HasFlag(TVTProgrammeDataFlag.CUSTOMPRODUCTION)
	local production:TProduction = TProduction(GetProductionCollection().GetByID())
	if production and production.productionConcept and production.productionConcept.script
		'hier dann ueberpruefen ob komplett oder nicht
	endif
endif
```

Reply:
Wie käme man denn an das Drehbuch, das einem Programm zugrunde liegt?
Aber wenn wir Refill für Serien unterbinden (nicht unterstützen) würde ein Kommentar im Code, der auf diesen Umstand hinweist, ausreichen

Reply:
Siehe dem Kommentar ueber Dir (hat sich wohl ueberschnitten).

Reply:
Der Code direkt verwendet führt erstmal zu einer zyklischen Abhängigkeit in TProgrammeLicence. Ich versuche mal, das an eine sinnvolle Stelle auszulagern.

Reply:
Und die GetProductionCollectio() gibt es auch (noch nicht + in derselben Datei wie TProduction, wäre also auch zyklisch)... Das sieht mir nach einer Ronny-Zuarbeit aus.

Reply:
Gegebenenfalls ließe sich hier das tradeable-Flag verwenden: eine Serie ist so lange nicht verkäuflich, bis sie komplett abgedreht ist. Sie muss zu dem Zeitpunkt verkäuflich werden, wo das ggf. noch nicht abgedrehte Script weggeworfen wird.

Reply:
> Der Code direkt verwendet führt erstmal zu einer zyklischen Abhängigkeit in TProgrammeLicence. Ich versuche mal, das an eine sinnvolle Stelle auszulagern.

Da hast Du recht - und es ist wohl auch "logisch" - denn einer Programmlizenz ist es erstmal egal, wie das zugrundeliegende Programm produziert wurde.
Wenn es das wirklich alles wissen koennen muss, muessten wir TProductionBase einfuehren und dort alles von "TProgrammeLicence" unabhaengige reinpacken.

Der Vorteil waere dann, dass ein Programm (bzw deren Lizenz) dann Beziehungen wie "vom gleichen Produzenten" usw. nutzen koennte - oder halt auch schauen kann, ob das zugrundeliegende Drehbuch "vollstaendig verfilmt" worden ist.



Allerdings kommt ja von Dir auch der Vorschlag, das "tradeable"-Flag zu nutzen ... ich wuerde das etwas anders aufziehen: Manchmal will man eine Serie "einfach verkaufen", auch wenn noch Teile fehlen.
"sauberer" waere es aber folgend:
- Beim Archiv hervorheben, wenn eine Serie noch "in Produktion" ist (also Serienkopf.programmeData.productionID > 0 - und Drehbuch noch nicht abgeschlossen -> da kann man ja den obigen Code "nutzen") -> das waere aehnlich dem "im Programmplan"
- wird eine solche Serie TROTZDEM in den Koffer gelegt, werden alle etwaig laufenden Produktionen abgebrochen und Konzepte/Drehbuecher vernichtet (Pech fuer die Kuh Else bezueglich geleisteter Anzahlungen) -> bei "Im Programmplan"-Lizenzen sorgt dies ja bereits fuer Sendeausfall, hat also auch Konsequenzen



Generell sind ja solche "Abfragen" immer mit CPU-Kosten versehen - da waere das Einfuehren einer ProgrammeData-Flag ("IS_COMPLETE_SERIES") durchaus in Erwaegung zu ziehen. Natuerlich nur, wenn man den Verkauf bei "nicht kompletten Serien" unterbinden woellte.

Auch koennten ProgrammeData noch "scriptID:Int" zugewiesen bekommen. Oder man ueberlegt, ob es bei der Menge an moeglichen IDs einen eigenen Container dafuer anzulegen - also sowas wie einen "extras:TContainer" in dem dann scriptID, producerID, .... usw eingelagert werden koennten.

Reply:
Ein "schmutziger" Ansatz, ist das nutzen des "GameInformation"-Systems. Die Quellcode-Dateien koennen ja selbst Funktionen registrieren, die dann per `GetGameInformation()` abgerufen werden koennten. Sprich eine "Anonyme"-Anbindung / entkoppelt.

```
GetGameInformationCollection().AddProvider("production", TProductionInformationProvider.GetInstance()) 'existiert noch nicht

'muesste von hier erben
Type TGameInformationProvider
	Method Set(key:string, obj:object) abstract

	Method Get:object(key:string, params:object = null, useTime:Long = 0) abstract

	Method GetString:string(key:string, params:object = null)
		return string(Get(key, params))
	End Method

	Method GetFloat:Float(key:string, params:object = null)
		return float(string(Get(key, params)))
	End Method

	Method GetInt:Int(key:string, params:object = null)
		return int(string(Get(key, params)))
	End Method

	Method GetLong:Long(key:string, params:object = null)
		return Long(string(Get(key, params)))
	End Method
End Type
```

Damit waere dann sowas moeglich: `GetGameInformation('production.isComplete", string(productionID)` oder halt alles was von uns von Interesse ist. Gleiches gaenge auch fuer Lizenzen, ProgrammeData. Damit koennen Informationen abgerufen werden, ohne direkte Abhaengigkeiten zu haben. Aehnlich dem Event-System.

Reply:
Ich würde den ganzen Ablauf mit dem Abbruch der Produktion, wenn man eine teilweise abgedrehte Serie in den Koffer legt, nicht verkomplizieren wollen.
Wenn man die Serie loswerden will, ist es ja noch mehr "Pech", wenn man den aktuellen Dreh fertigstellen muss und die Folge dann auch nochmal senden...

Zum Unterschied bei der CPU-Belastung (tradeable vs. neues Flag) kann ich nicht so viel sagen.

Reply:
Ich empfinde es irgendwie "unlogisch", dass ich ein Drehbuch wegwerfen muss um eine (unfertige) Serie verkaufen zu koennen - das passiert auch in zwei unterschiedlichen Raeumen.

Wenn ein Film im Programmplan ist, kann ich ja trotzdem verkaufen.
Also wuerde ich es auch so bei den unfertigen Ketten handhaben wollen: 
- Aufgelistet im Archiv
- Hinweis im Datenblatt auf "In Produktion!!"
- Abbruch der Produktionen bei "im Koffer" und "Archiv verlassen"
- Nicht ausgestrahlte Kettenelemente haben vlt keinen "richtigen" Verkaufspreis - Pech!

Reply:
Ich empfinde es irgendwie "unlogisch", dass ich ein Drehbuch wegwerfen muss um eine (unfertige) Serie verkaufen zu koennen - das passiert auch in zwei unterschiedlichen Raeumen.

Wenn ein Film im Programmplan ist, kann ich ja trotzdem verkaufen.
Also wuerde ich es auch so bei den unfertigen Ketten handhaben wollen: 
- Aufgelistet im Archiv
- Hinweis im Datenblatt auf "In Produktion!!"
- Abbruch der Produktionen bei "im Koffer" und "Archiv verlassen"
- Nicht ausgestrahlte Kettenelemente haben vlt keinen "richtigen" Verkaufspreis - Pech!

Reply:
Dein derzeitiger Ansatz ("tradeable") sollte damit ja kompatibel sein - oder nicht? Man muesste ihn dann nach Umsetzung nur wieder rausnehmen.

Reply:
Für mich ist es nicht so unlogisch. Ein Film im Programmplan ist ja schon fertig und kann deshalb grundsätzlich verkauft werden. Eine nicht abgedrehte Serie ist nicht fertig - solange das Drehbuch noch Folgen, die gedreht werden können, enthält, kann ich noch nicht verkaufen. Daher muss ich mich entscheiden - Drehbuch wegwerfen oder fertigdrehen.